### PR TITLE
IC-1746: Transformed Desired Outcomes to accept selected service cate…

### PR DESCRIPTION
**Pull request by andrewburrowsmoj**
…gory

## What does this pull request do?

Provide ability to set desired outcome on a draft referral for a given service category

## What is the intent behind these changes?

To provide SP users the ability to provide desired outcomes for cohort referrals.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:166]: I'm assuming here that we can allow the caller to remove all desired outcomes for a service category. 

I assume this will be required if the user deselects the service category from the make a referral - we will also need to remove any desired outcomes as well.

However the behaviour before was to prevent it being unset... there could be a check in the ReferralService to ensure that at least one desired outcome for all service categories is left. Do you think I should add this check?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt:30]: Realise that I should be using a new DTO object rather than Map based on the code in the UI.
Going to make that change quickly just now...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:170]: we need to explicitly stop unsetting of desiredoutcomes. previously this check handled this in `validateDraftReferralUpdate`

```
      if (it.isEmpty()) {
        errors.add(FieldError(field = "desiredOutcomesIds", error = Code.CANNOT_BE_EMPTY))
      }
      ```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt:65]: please can this go in it's own file

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt:65]: it would be fine here, except that it's also used on its own outside of the context of DraftReferralDTO, which ReferralComplexityLevel is not

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:170]: (talking to myself here)

.. we should make sure that we remove desiredOutcomes and complexityLevel for a given serviceCategoryId when it is removed via the updateDraftReferral route.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:309]: nit: you shouldn't need the not null assertion here. the compiler will pick up that it cannot be null due to the above check ( i think)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:298]: variable name copy paste error

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt:65]: Ok, I did originally, but noticed that InterventionDTO had a DTO in the same file.
I'll do that change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:309]: i'm not sure using the referential equality operator is correct here (`===`). (too much javascript ?? 😛 )

Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:309]: Oh... my IDE makes it red if I don't have ? or !!

Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:309]: hmm somehow it appears to work on the unit test, but I'll update to '=='

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:170]: not following the comment about removing them from updateDraftReferral.. aren't they already removed in this PR here https://github.com/ministryofjustice/hmpps-interventions-service/pull/253/files#diff-f7a6efe2090c4b294417026b860873e6c0fbc3f1b603412b45a294f1e44150c9L250 ?

i think we just need to check if the list of desiredoutcomes is empty in the new referral service method for updating the outcomes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt]: fixed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt]: fixed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt]: fixed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:170]: Added back in the check for empty list. 

Setting to empty list is still possible via UpdateDraftReferral when unselecting a service Category. This will be a new PR with a new ticket (still to be created).

**Comment by andrewburrowsmoj** [comment]
Rebased from main and resolve conflicts

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:308]: small thing: these `FieldError` objects are intended to be used when there are multiple fields being updating in a request. in this case, because we are only ever updating desired outcomes, i think just a plain 400 error is preferable here. it is more consistent with the responses generated for other error conditions below.

File: src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql
Comment (tomsmyers) [inline src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql:1]: just noticed this lonely `begin`, what does it do?

Comment (tomsmyers) [reply src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql:1]: i guess it goes with the `commit` at the end?

**Review by tomsmyers — APPROVED** [review]
approving now but would be good to get the error code thing changed before merge. thanks!

File: src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql
Comment (andrewburrowsmoj) [reply src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql:1]: It wraps up the whole set of statements into a single transaction so rolls back if for instance the update fails.

It was more for when I was running directly on the database for testing - here it's not really needed but shouldn't hurt to have it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt]: Ok. I assumed because it's something that is directly related to the user's parameters (rather than indirectly) that we needed to use this type of exception.

I have changed the exception to ServerWebInputException

File: src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql
Comment (tomsmyers) [reply src/main/resources/db/migration/V1_53__referral_selected_desired_outcomes.sql:1]: nice

**Pull request merged**
